### ISSUE TITLE
[Bug]: [Nightly] Auto compaction sometimes runs slowly

### ISSUE THREAD

**Issue by binbinlv** [milvus-io/milvus#13045]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: latest
- Deployment mode(standalone or cluster): standalone & cluster
- SDK version(e.g. pymilvus v2.0.0rc2): pymilvus rc9.22
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

Auto compaction (segments reached 10), compaction could not be completed in 60s

```
[2021-12-09T00:38:55.908Z] >               raise BaseException(1, "Ccompact auto-merge more than 60s")
[2021-12-09T00:38:55.908Z] E               BaseException: (1, 'Ccompact auto-merge more than 60s')
```

### Expected Behavior

Auto compaction (segments reached 10), compaction could be completed in relatively short time. 

### Steps To Reproduce

Run nightly:
https://ci.milvus.io:18080/jenkins/blue/organizations/jenkins/milvus-nightly-ci/detail/master/256/pipeline/106

Logs:
https://ci.milvus.io:18080/jenkins/blue/organizations/jenkins/milvus-nightly-ci/detail/master/256/artifacts

Failed case:
testcases/test_compaction.py::TestCompactionOperation::test_compact_threshold_auto_merge

Failed collection name:
compact_QmYxDTZS

TimeLine:

```
[2021-12-08T22:07:42.854Z] [gw2] [ 17%] FAILED testcases/test_compaction.py::TestCompactionOperation::test_compact_threshold_auto_merge 
```

**Comment by XuanYang-cn** [comment]
Seems like `auto-compaction` is off. Please turn the `auto-compaction` on and test again.

/assign @binbinlv
/unassign

**Comment by yanliang567** [comment]
/assign @ThreadDao 
please help to update the latest status of these tests
/unassign @binbinlv

**Comment by ThreadDao** [comment]
Now `auto-compaction` is on and the problem did not reappear.

**Comment by binbinlv** [comment]
This issue appears again, so reopen it.
Nightly: [standalone: artifacts-milvus-standalone-master-276-pymilvus-e2e-logs.tar.gz]
https://ci.milvus.io:18080/jenkins/blue/organizations/jenkins/milvus-nightly-ci/detail/master/276/artifacts

Test logs: [artifacts-milvus-standalone-pymilvus-pytest-logs.tar.gz]
https://ci.milvus.io:18080/jenkins/blue/organizations/jenkins/milvus-nightly-ci/detail/master/276/artifacts

Failed case:
test_compact_threshold_auto_merge
```
>               raise BaseException(1, "Compact auto-merge more than 60s")
E               BaseException: (1, 'Compact auto-merge more than 60s')
```

Failed collection name:
compact_0O3zxHRH

Timeline: [2021-12-24T15:40:40.730Z] testcases/test_compaction.py::TestCompactionOperation::test_compact_threshold_auto_merge FAILED

**Comment by XuanYang-cn** [comment]
The reason is:

`Load` arrived before auto-compaction finished. So querynode loaded the segments before merging.
```
1829929:[2021/12/24 15:39:39.370 +00:00] [DEBUG] [task.go:426] ["loadCollectionTask: add a childTask"] 
```

And `handoff` failed because handoff happened during load
```
1829941:[2021/12/24 15:39:39.371 +00:00] [DEBUG] [query_coord.go:431] ["watchHandoffSegmentLoop: collection/partition has not been 
```

**Comment by XuanYang-cn** [comment]
The case:
```python
    def test_compact_threshold_auto_merge(self):
        """
        target: test num (segment_size < 1/2Max) reaches auto-merge threshold 10
        method: 1.create with shard_num=1
                2.insert flush 10 times (merge threshold 10)
                3.wait for compaction, load
        expected: Get query segments info to verify segments auto-merged into one
        """
        threshold = ct.compact_segment_num_threshold
    
        # create collection shard_num=1, insert 10 segments, each with one entity
        collection_w = self.collection_insert_multi_segments_one_shard(prefix, num_of_segment=threshold)
    
        # Estimated auto-merging takes 30s
        cost = 60
        collection_w.load()
        start = time()
        while True:
            sleep(5)
            segments_info = self.utility_wrap.get_query_segment_info(collection_w.name)[0]
    
            # verify segments reaches threshold, auto-merge ten segments into one
            if len(segments_info) == 1:
                break
            end = time()
            if end - start > cost:
>               raise BaseException(1, "Compact auto-merge more than 60s")
E               BaseException: (1, 'Compact auto-merge more than 60s')

```

**Comment by xiaofan-luan** [comment]
@xige-16 is this a duplicated issue?

**Comment by binbinlv** [comment]
It occurs again in the following nightly:
https://ci.milvus.io:18080/jenkins/blue/organizations/jenkins/weekly-milvus-nightly-ci%2Fweekly-milvus-ci-test/detail/weekly-milvus-ci-test/48/pipeline

1. Test log: [standalone: [milvus-standalone-48-pymilvus-e2e-logs.tar.gz]
 https://ci.milvus.io:18080/jenkins/job/weekly-milvus-nightly-ci/job/weekly-milvus-ci-test/48/artifact/milvus-standalone-48-pymilvus-e2e-logs.tar.gz

2. Test case fail time:
[2022-02-25T13:51:58.128Z] [gw2] [ 23%] FAILED testcases/test_compaction.py::TestCompactionOperation::test_compact_threshold_auto_merge 

3. Test case:
```
[2022-02-25T14:44:41.162Z] __________ TestCompactionOperation.test_compact_threshold_auto_merge ___________
[2022-02-25T14:44:41.162Z] [gw2] linux -- Python 3.6.8 /usr/local/bin/python3
[2022-02-25T14:44:41.162Z] 
[2022-02-25T14:44:41.162Z] self = <test_compaction.TestCompactionOperation object at 0x7fb6ddd5f7f0>
[2022-02-25T14:44:41.162Z] 
[2022-02-25T14:44:41.162Z]     @pytest.mark.tags(CaseLabel.L2)
[2022-02-25T14:44:41.162Z]     def test_compact_threshold_auto_merge(self):
[2022-02-25T14:44:41.162Z]         """
[2022-02-25T14:44:41.162Z]         target: test num (segment_size < 1/2Max) reaches auto-merge threshold 10
[2022-02-25T14:44:41.162Z]         method: 1.create with shard_num=1
[2022-02-25T14:44:41.162Z]                 2.insert flush 10 times (merge threshold 10)
[2022-02-25T14:44:41.162Z]                 3.wait for compaction, load
[2022-02-25T14:44:41.162Z]         expected: Get query segments info to verify segments auto-merged into one
[2022-02-25T14:44:41.162Z]         """
[2022-02-25T14:44:41.162Z]         threshold = ct.compact_segment_num_threshold
[2022-02-25T14:44:41.162Z]     
[2022-02-25T14:44:41.162Z]         # create collection shard_num=1, insert 10 segments, each with one entity
[2022-02-25T14:44:41.162Z]         collection_w = self.collection_insert_multi_segments_one_shard(prefix, num_of_segment=threshold)
[2022-02-25T14:44:41.162Z]     
[2022-02-25T14:44:41.162Z]         # Estimated auto-merging takes 30s
[2022-02-25T14:44:41.162Z]         cost = 60
[2022-02-25T14:44:41.162Z]         collection_w.load()
[2022-02-25T14:44:41.162Z]         start = time()
[2022-02-25T14:44:41.162Z]         while True:
[2022-02-25T14:44:41.162Z]             sleep(5)
[2022-02-25T14:44:41.162Z]             segments_info = self.utility_wrap.get_query_segment_info(collection_w.name)[0]
[2022-02-25T14:44:41.162Z]     
[2022-02-25T14:44:41.162Z]             # verify segments reaches threshold, auto-merge ten segments into one
[2022-02-25T14:44:41.162Z]             if len(segments_info) == 1:
[2022-02-25T14:44:41.162Z]                 break
[2022-02-25T14:44:41.162Z]             end = time()
[2022-02-25T14:44:41.162Z]             if end - start > cost:
[2022-02-25T14:44:41.162Z] >               raise BaseException(1, "Compact auto-merge more than 60s")
[2022-02-25T14:44:41.162Z] E               BaseException: (1, 'Compact auto-merge more than 60s')
```

**Comment by binbinlv** [comment]
@xige-16 Could you help to have a look at this issue, thanks.

**Comment by binbinlv** [comment]
It occurs again in the following nightly:
https://ci.milvus.io:18080/jenkins/blue/organizations/jenkins/milvus-nightly-ci/detail/master/372/pipeline/106

1.log
[artifacts-milvus-standalone-master-372-pymilvus-e2e-logs.tar.gz](https://ci.milvus.io:18080/jenkins/job/milvus-nightly-ci/job/master/372/artifact/artifacts-milvus-standalone-master-372-pymilvus-e2e-logs.tar.gz)

2. Test case fail time:
[2022-02-28T19:03:02.582Z] [gw2] [ 22%] FAILED testcases/test_compaction.py::TestCompactionOperation::test_compact_threshold_auto_merge 

3. collection name: compact_OkYOJVuV

4. Test case:
```python
       """
      target: test num (segment_size < 1/2Max) reaches auto-merge threshold 10
      method: 1.create with shard_num=1
                     2.insert flush 10 times (merge threshold 10)
                     3.wait for compaction, load
      expected: Get query segments info to verify segments auto-merged into one
      """
     threshold = ct.compact_segment_num_threshold
  
     # create collection shard_num=1, insert 10 segments, each with one entity
     collection_w = self.collection_insert_multi_segments_one_shard(prefix, num_of_segment=threshold)

     # Estimated auto-merging takes 30s
     cost = 60
     collection_w.load()
     start = time()
     while True:
     sleep(5)
     segments_info = self.utility_wrap.get_query_segment_info(collection_w.name)[0]

      # verify segments reaches threshold, auto-merge ten segments into one
      if len(segments_info) == 1:
      break
      end = time()
      if end - start > cost:
      raise BaseException(1, "Compact auto-merge more than 60s")
      BaseException: (1, 'Compact auto-merge more than 60s')
```

**Comment by yanliang567** [comment]
@jaime0815 could you please take a look at this issue? It reproduces frequently these days.
/assign @jaime0815 
/unassign

**Comment by xige-16** [comment]
> The reason is:
> 
> `Load` arrived before auto-compaction finished. So querynode loaded the segments before merging.
> 
> ```
> 1829791:[2021/12/24 15:39:39.347 +00:00] [DEBUG] [task.go:333] ["loadCollectionTask: get collection's all partitionIDs"]
> ```
> 
> ```
> 1829899:[2021/12/24 15:39:39.369 +00:00] [DEBUG] [services.go:747] ["success to complete compaction"] [planID=430010631267024898]
> ```
> 
> And `handoff` failed because handoff happened during load
> 
> ```
> 1829941:[2021/12/24 15:39:39.371 +00:00] [DEBUG] [query_coord.go:431] ["watchHandoffSegmentLoop: collection/partition has not been loaded or autoHandoff equal to false, remove req from etcd"]
> ```

querycoord add collection info to meta cache when load collection done, before that the compaction has been completed and querycoord is notified, but when querycoord checks meta, it is found that the collection meta does not exist in the querycoord memory, and the segment after compaction is not loaded into the querynode.

I'll fix it soon.

**Comment by binbinlv** [comment]
It occurs again in the following nightly:
1. Logs:
[milvus-distributed-75-pymilvus-e2e-logs.tar.gz](https://ci.milvus.io:18080/jenkins/job/weekly-milvus-nightly-ci/job/weekly-milvus-ci-test/75/artifact/milvus-distributed-75-pymilvus-e2e-logs.tar.gz)
2. Test failed time: 
[2022-03-06T06:31:44.414Z] [gw3] [ 23%] FAILED testcases/test_compaction.py::TestCompactionOperation::test_compact_threshold_auto_merge 
3. collection name: compact_L5HV6of5

**Comment by xiaofan-luan** [comment]
> > The reason is:
> > `Load` arrived before auto-compaction finished. So querynode loaded the segments before merging.
> > ```
> > 1829791:[2021/12/24 15:39:39.347 +00:00] [DEBUG] [task.go:333] ["loadCollectionTask: get collection's all partitionIDs"]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > 1829899:[2021/12/24 15:39:39.369 +00:00] [DEBUG] [services.go:747] ["success to complete compaction"] [planID=430010631267024898]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And `handoff` failed because handoff happened during load
> > ```
> > 1829941:[2021/12/24 15:39:39.371 +00:00] [DEBUG] [query_coord.go:431] ["watchHandoffSegmentLoop: collection/partition has not been loaded or autoHandoff equal to false, remove req from etcd"]
> > ```
> 
> querycoord add collection info to meta cache when load collection done, before that the compaction has been completed and querycoord is notified, but when querycoord checks meta, it is found that the collection meta does not exist in the querycoord memory, and the segment after compaction is not loaded into the querynode.
> 
> I'll fix it soon.

We can not apply patch by patch to fix the issues one by one. 
For example, if load operation start, the collection should be in meta already with a loading state.
Once the load finish, the collection state can be changed to "loaded".
I'm really worried we patched too many to the logic and make it hard to maintain(Shouldn't rely on ordering because in a distributed system one can not consider all the concurrent cases)

**Comment by binbinlv** [comment]
The test case "test_compact_after_binary_index" failed for "BaseException: (1, 'Handoff after compact and index cost more than 1s')" in the following nightly:

https://ci.milvus.io:18080/jenkins/blue/organizations/jenkins/milvus-nightly-ci/detail/master/386/pipeline/

I think the root cause is the same with this issue, @xige-16  could you help to confirm? Thanks.

1 Log:
[artifacts-milvus-distributed-master-386-pymilvus-e2e-logs.tar.gz](https://ci.milvus.io:18080/jenkins/job/milvus-nightly-ci/job/master/386/artifact/artifacts-milvus-distributed-master-386-pymilvus-e2e-logs.tar.gz)

2 collection name: "compact_igouk2W2"

3Test case failed time:
[2022-03-07T19:03:52.610Z] [gw0] [ 12%] FAILED testcases/test_compaction.py::TestCompactionOperation::test_compact_after_binary_index 

4 Test case:
```
    def test_compact_after_binary_index(self):
        """
        target: test compact after create index
        method: 1.insert binary data into two segments
                2.create binary index
                3.compact
                4.search
        expected: Verify segment info and index info
        """
        # create collection with 1 shard and insert 2 segments
        collection_w = self.init_collection_wrap(name=cf.gen_unique_str(prefix), shards_num=1,
                                                 schema=cf.gen_default_binary_collection_schema())
        for i in range(2):
            df, _ = cf.gen_default_binary_dataframe_data()
            collection_w.insert(data=df)
            assert collection_w.num_entities == (i + 1) * ct.default_nb

        # create index
        collection_w.create_index(ct.default_binary_vec_field_name, ct.default_binary_index)
        log.debug(collection_w.index())

        # load and search
        collection_w.load()
        search_params = {"metric_type": "JACCARD", "params": {"nprobe": 32}}
        search_res_one, _ = collection_w.search(df[ct.default_binary_vec_field_name][:ct.default_nq].to_list(),
                                                ct.default_binary_vec_field_name, search_params, ct.default_limit)

        # compact
        collection_w.compact()
        collection_w.wait_for_compaction_completed()
        c_plans = collection_w.get_compaction_plans(check_task=CheckTasks.check_merge_compact)[0]

        # waiting for handoff completed and search
        cost = 1
        start = time()
        while True:
            sleep(5)
            segment_info = self.utility_wrap.get_query_segment_info(collection_w.name)[0]
            if len(segment_info) != 0 and segment_info[0].segmentID == c_plans.plans[0].target:
                log.debug(segment_info)
                break
            if time() - start > cost:
                raise BaseException(1, f"Handoff after compact and index cost more than {cost}s")
```

**Comment by congqixia** [comment]
Now the test case waits handoff complete for only 1 second. It's reasonable when there is only when test case running. While in CI env, there will be 6 workers running different cases. Since hand-off operation may be blocked by query/search task put into QueryChannel before, it has to wait until the Search/Query operation is done.
Maybe we shall adjust this timeout value to a more reasonable value. @binbinlv @ThreadDao

**Comment by binbinlv** [comment]
Auto compaction slowly again in the following nightly tests:
https://ci.milvus.io:18080/jenkins/blue/organizations/jenkins/milvus-nightly-ci/detail/master/398/pipeline

1.log
[artifacts-milvus-standalone-master-398-pymilvus-e2e-logs.tar.gz](https://ci.milvus.io:18080/jenkins/job/milvus-nightly-ci/job/master/398/artifact/artifacts-milvus-standalone-master-398-pymilvus-e2e-logs.tar.gz)

2 Test case fail time:
[2022-03-13T19:03:10.131Z] [gw2] [ 23%] FAILED testcases/test_compaction.py::TestCompactionOperation::test_compact_threshold_auto_merge 

3 collection name:
compact_bf6sM5XX

**Comment by binbinlv** [comment]
> > > The reason is:
> > > `Load` arrived before auto-compaction finished. So querynode loaded the segments before merging.
> > > ```
> > > 1829791:[2021/12/24 15:39:39.347 +00:00] [DEBUG] [task.go:333] ["loadCollectionTask: get collection's all partitionIDs"]
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > ```
> > > 1829899:[2021/12/24 15:39:39.369 +00:00] [DEBUG] [services.go:747] ["success to complete compaction"] [planID=430010631267024898]
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > And `handoff` failed because handoff happened during load
> > > ```
> > > 1829941:[2021/12/24 15:39:39.371 +00:00] [DEBUG] [query_coord.go:431] ["watchHandoffSegmentLoop: collection/partition has not been loaded or autoHandoff equal to false, remove req from etcd"]
> > > ```
> > 
> > 
> > querycoord add collection info to meta cache when load collection done, before that the compaction has been completed and querycoord is notified, but when querycoord checks meta, it is found that the collection meta does not exist in the querycoord memory, and the segment after compaction is not loaded into the querynode.
> > I'll fix it soon.
> 
> We can not apply patch by patch to fix the issues one by one. For example, if load operation start, the collection should be in meta already with a loading state. Once the load finish, the collection state can be changed to "loaded". I'm really worried we patched too many to the logic and make it hard to maintain(Shouldn't rely on ordering because in a distributed system one can not consider all the concurrent cases)

@xiaofan-luan @xige-16 
so what is the new solution here?

**Comment by xiaofan-luan** [comment]
@xige-16 we should fix this by adding a lease on datacoord side to maintain data existence when compaction happened.

**Comment by xige-16** [comment]
releated with #15662

**Comment by xiaofan-luan** [comment]
@xige-16  so what is the expected behaviour?

**Comment by binbinlv** [comment]
Keep open for verification. (this did not merge any fix, and will verify whether it has been reproduced recently.

**Comment by binbinlv** [comment]
It has not been reproduced for a period of time, so close it first and will reopen it if it occurs in the future.

**Comment by binbinlv** [comment]
It is reproduced in the latest nightly, so reopen it.

**Comment by binbinlv** [comment]
It is reproduced in both pulsar standalone and kafka cluster pipeline.

pulsar standalone:

https://ci.milvus.io:18080/jenkins/blue/organizations/jenkins/milvus-nightly-ci/detail/master/605/pipeline/124/

1. milvus log:
[artifacts-milvus-standalone-master-605-pymilvus-e2e-logs.tar.gz](https://ci.milvus.io:18080/jenkins/job/milvus-nightly-ci/job/master/605/artifact/artifacts-milvus-standalone-master-605-pymilvus-e2e-logs.tar.gz)

2. collection name: compact_GBKMKrG4

kafka cluster:
https://ci.milvus.io:18080/jenkins/blue/rest/organizations/jenkins/pipelines/milvus-nightly-ci/branches/master/runs/605/nodes/141/steps/178/log/?start=0

1. milvus log
[artifacts-milvus-distributed-kafka-master-605-pymilvus-e2e-logs.tar.gz](https://ci.milvus.io:18080/jenkins/job/milvus-nightly-ci/job/master/605/artifact/artifacts-milvus-distributed-kafka-master-605-pymilvus-e2e-logs.tar.gz)

2. collection name: compact_33HWg8R0

**Comment by binbinlv** [comment]
Same issue in the latest nightly:
https://ci.milvus.io:18080/jenkins/blue/organizations/jenkins/milvus-nightly-ci/detail/master/608/pipeline/190
1. milvus log
[artifacts-milvus-distributed-pulsar-master-608-pymilvus-e2e-logs.tar.gz](https://ci.milvus.io:18080/jenkins/job/milvus-nightly-ci/job/master/608/artifact/artifacts-milvus-distributed-pulsar-master-608-pymilvus-e2e-logs.tar.gz)

2. collection name: compact_nPTANFQ1

**Comment by binbinlv** [comment]
Reproduced in high probability, so set it as urgent.

**Comment by xiaofan-luan** [comment]
should be fixed by #17689

**Comment by binbinlv** [comment]
Did not occur last nightly, cancel urgent first, and will close it if it it is not reproduced for several days.

**Comment by binbinlv** [comment]
It occurs again in the latest nightly:
https://ci.milvus.io:18080/jenkins/blue/organizations/jenkins/milvus-nightly-ci/detail/master/612/pipeline/190

1. Milvus logs:
[artifacts-milvus-distributed-pulsar-master-612-pymilvus-e2e-logs.tar.gz](https://ci.milvus.io:18080/jenkins/job/milvus-nightly-ci/job/master/612/artifact/artifacts-milvus-distributed-pulsar-master-612-pymilvus-e2e-logs.tar.gz)

2. collection name: compact_oNDgY3kw

3. Test cases:
```

[2022-06-23T19:54:14.107Z]     @pytest.mark.tags(CaseLabel.L2)

[2022-06-23T19:54:14.107Z]     def test_compact_threshold_auto_merge(self):

[2022-06-23T19:54:14.107Z]         """

[2022-06-23T19:54:14.107Z]         target: test num (segment_size < 1/2Max) reaches auto-merge threshold 10

[2022-06-23T19:54:14.107Z]         method: 1.create with shard_num=1

[2022-06-23T19:54:14.107Z]                 2.insert flush 10 times (merge threshold 10)

[2022-06-23T19:54:14.107Z]                 3.wait for compaction, load

[2022-06-23T19:54:14.107Z]         expected: Get query segments info to verify segments auto-merged into one

[2022-06-23T19:54:14.107Z]         """

[2022-06-23T19:54:14.107Z]         threshold = ct.compact_segment_num_threshold

[2022-06-23T19:54:14.107Z]     

[2022-06-23T19:54:14.107Z]         # create collection shard_num=1, insert 10 segments, each with one entity

[2022-06-23T19:54:14.107Z]         collection_w = self.collection_insert_multi_segments_one_shard(prefix, num_of_segment=threshold)

[2022-06-23T19:54:14.107Z]     

[2022-06-23T19:54:14.107Z]         # Estimated auto-merging takes 30s

[2022-06-23T19:54:14.107Z]         cost = 60

[2022-06-23T19:54:14.107Z]         collection_w.load()

[2022-06-23T19:54:14.107Z]         replicas = collection_w.get_replicas()[0]

[2022-06-23T19:54:14.107Z]         replica_num = len(replicas.groups)

[2022-06-23T19:54:14.107Z]         start = time()

[2022-06-23T19:54:14.107Z]         while True:

[2022-06-23T19:54:14.107Z]             sleep(5)

[2022-06-23T19:54:14.107Z]             segments_info = self.utility_wrap.get_query_segment_info(collection_w.name)[0]

[2022-06-23T19:54:14.107Z]     

[2022-06-23T19:54:14.107Z]             # verify segments reaches threshold, auto-merge ten segments into one

[2022-06-23T19:54:14.107Z]             if len(segments_info) == 1*replica_num:

[2022-06-23T19:54:14.107Z]                 break

[2022-06-23T19:54:14.107Z]             end = time()

[2022-06-23T19:54:14.107Z]             if end - start > cost:

[2022-06-23T19:54:14.107Z] >               raise MilvusException(1, "Compact auto-merge more than 60s")

[2022-06-23T19:54:14.107Z] E               pymilvus.exceptions.MilvusException: <MilvusException: (code=1, message=Compact auto-merge more than 60s)>
```

**Comment by binbinlv** [comment]
It occurs again in the latest nightly:

[artifacts-milvus-distributed-pulsar-master-618-pymilvus-e2e-logs.tar.gz](https://ci.milvus.io:18080/jenkins/job/milvus-nightly-ci/job/master/618/artifact/artifacts-milvus-distributed-pulsar-master-618-pymilvus-e2e-logs.tar.gz)

Collection: compact_QuVI9OWy

**Comment by binbinlv** [comment]
It occurs in the following nightly:
[artifacts-milvus-standalone-master-617-pymilvus-e2e-logs.tar.gz](https://ci.milvus.io:18080/jenkins/job/milvus-nightly-ci/job/master/617/artifact/artifacts-milvus-standalone-master-617-pymilvus-e2e-logs.tar.gz)

Collection: compact_2IipCD75

**Comment by binbinlv** [comment]
It occurs in the following latest nightly:
https://ci.milvus.io:18080/jenkins/blue/organizations/jenkins/milvus-nightly-ci/detail/master/620/pipeline

(1) milvus logs:
[artifacts-milvus-standalone-master-620-pymilvus-e2e-logs.tar.gz](https://ci.milvus.io:18080/jenkins/job/milvus-nightly-ci/job/master/620/artifact/artifacts-milvus-standalone-master-620-pymilvus-e2e-logs.tar.gz)
(2) collection name: compact_kcnuW7Qd

**Comment by binbinlv** [comment]
Add compaction time to 120s, it still exists. It occurs in the following latest nightly:
https://ci.milvus.io:18080/jenkins/blue/organizations/jenkins/milvus-nightly-ci/detail/master/634/pipeline/157/

(1) milvus logs: [artifacts-milvus-distributed-kafka-master-634-pymilvus-e2e-logs.tar.gz](https://ci.milvus.io:18080/jenkins/job/milvus-nightly-ci/job/master/634/artifact/artifacts-milvus-distributed-kafka-master-634-pymilvus-e2e-logs.tar.gz)
(2) collection name: compact_ji6fhq1Y

**Comment by binbinlv** [comment]
Same issue in latest nightly:
1. Nightly: https://ci.milvus.io:18080/jenkins/blue/rest/organizations/jenkins/pipelines/milvus-nightly-ci/branches/master/runs/646/nodes/124/steps/162/log/?start=0
2. collection name: compact_06xXdasR
3. FAILED time: [2022-07-10T19:01:55.460Z] [gw0] [ 18%] FAILED testcases/test_compaction.py::TestCompactionOperation::test_compact_threshold_auto_merge

**Comment by binbinlv** [comment]
Same issue in latest nightly:
https://ci.milvus.io:18080/jenkins/blue/organizations/jenkins/milvus-nightly-ci/detail/master/647/pipeline/190

1. milvus log:[artifacts-milvus-distributed-kafka-master-647-pymilvus-e2e-logs.tar.gz](https://ci.milvus.io:18080/jenkins/job/milvus-nightly-ci/job/master/647/artifact/artifacts-milvus-distributed-kafka-master-647-pymilvus-e2e-logs.tar.gz)
2. collection name: compact_abb9iPVJ
3. Failed time: [2022-07-11T15:10:44.236Z] [gw5] [ 19%] FAILED testcases/test_compaction.py::TestCompactionOperation::test_compact_threshold_auto_merge

**Comment by binbinlv** [comment]
Continuously appears in latest nightly:
https://jenkins.milvus.io:18080/blue/organizations/jenkins/Milvus%20Nightly%20CI/detail/master/16/pipeline

1. milvus log:[artifacts-milvus-distributed-pulsar-nightly-16-pymilvus-e2e-logs.tar.gz](https://jenkins.milvus.io:18080/job/Milvus%20Nightly%20CI/job/master/16/artifact/artifacts-milvus-distributed-pulsar-nightly-16-pymilvus-e2e-logs.tar.gz)
2. collection name: compact_Bxo9zdWZ
3. FAILED time:
[2022-08-04T19:07:53.810Z] [gw2] [ 20%] FAILED testcases/test_compaction.py::TestCompactionOperation::test_compact_threshold_auto_merge

**Comment by wayblink** [comment]
> /assign @wayblink pls help on it

Key Error:
2022-08-05T03:05:52.735735091+08:00 stdout F [2022/08/04 19:05:52.735 +00:00] [ERROR] [datacoord/services.go:817] ["failed to complete compaction"] [planID=435064655754297345] [error="can not be compacted, segment with ID 435064652949094401 has reference lock"]

Test process:
1,create & insert
2,load
3,auto compaction by milvus
4,getQuerySegmentInfo, expect to get only one segment after compaction

Auto compaction process：
1,datacoord compaction_tigger generate compactionPlan
2,datacoord call a certain datanode to Compaction, set segments' state to Compacting
3,datanode execute Compaction
4,datanode call datacoord CompleteCompaction to report status and update segment meta(add new segment, drop old segemnts)

CompleteCompaction failed because there are segmentRefLocks required by LoadCollection.
After CompleteCompaction failed, datacoord doesn't update segments' state, still in Compacting state.
So they won't be triggered compaction again until timeout, 3*mins, which is longer than the test case threshold.

We will rethink and enhance the compaction process.

**Comment by XuanYang-cn** [comment]
> @xiaofan-luan please help to investigate again, thanks.
> 
> It occurs again in the latest nightly: https://ci.milvus.io:18080/jenkins/blue/organizations/jenkins/milvus-nightly-ci/detail/master/612/pipeline/190
> 
> 1. Milvus logs:
>    [artifacts-milvus-distributed-pulsar-master-612-pymilvus-e2e-logs.tar.gz](https://ci.milvus.io:18080/jenkins/job/milvus-nightly-ci/job/master/612/artifact/artifacts-milvus-distributed-pulsar-master-612-pymilvus-e2e-logs.tar.gz)
> 2. collection name: compact_oNDgY3kw
> 3. Test cases:
> 
> ```
> 
> [2022-06-23T19:54:14.107Z]     @pytest.mark.tags(CaseLabel.L2)
> 
> [2022-06-23T19:54:14.107Z]     def test_compact_threshold_auto_merge(self):
> 
> [2022-06-23T19:54:14.107Z]         """
> 
> [2022-06-23T19:54:14.107Z]         target: test num (segment_size < 1/2Max) reaches auto-merge threshold 10
> 
> [2022-06-23T19:54:14.107Z]         method: 1.create with shard_num=1
> 
> [2022-06-23T19:54:14.107Z]                 2.insert flush 10 times (merge threshold 10)
> 
> [2022-06-23T19:54:14.107Z]                 3.wait for compaction, load
> 
> [2022-06-23T19:54:14.107Z]         expected: Get query segments info to verify segments auto-merged into one
> 
> [2022-06-23T19:54:14.107Z]         """
> 
> [2022-06-23T19:54:14.107Z]         threshold = ct.compact_segment_num_threshold
> 
> [2022-06-23T19:54:14.107Z]     
> 
> [2022-06-23T19:54:14.107Z]         # create collection shard_num=1, insert 10 segments, each with one entity
> 
> [2022-06-23T19:54:14.107Z]         collection_w = self.collection_insert_multi_segments_one_shard(prefix, num_of_segment=threshold)
> 
> [2022-06-23T19:54:14.107Z]     
> 
> [2022-06-23T19:54:14.107Z]         # Estimated auto-merging takes 30s
> 
> [2022-06-23T19:54:14.107Z]         cost = 60
> 
> [2022-06-23T19:54:14.107Z]         collection_w.load()
> 
> [2022-06-23T19:54:14.107Z]         replicas = collection_w.get_replicas()[0]
> 
> [2022-06-23T19:54:14.107Z]         replica_num = len(replicas.groups)
> 
> [2022-06-23T19:54:14.107Z]         start = time()
> 
> [2022-06-23T19:54:14.107Z]         while True:
> 
> [2022-06-23T19:54:14.107Z]             sleep(5)
> 
> [2022-06-23T19:54:14.107Z]             segments_info = self.utility_wrap.get_query_segment_info(collection_w.name)[0]
> 
> [2022-06-23T19:54:14.107Z]     
> 
> [2022-06-23T19:54:14.107Z]             # verify segments reaches threshold, auto-merge ten segments into one
> 
> [2022-06-23T19:54:14.107Z]             if len(segments_info) == 1*replica_num:
> 
> [2022-06-23T19:54:14.107Z]                 break
> 
> [2022-06-23T19:54:14.107Z]             end = time()
> 
> [2022-06-23T19:54:14.107Z]             if end - start > cost:
> 
> [2022-06-23T19:54:14.107Z] >               raise MilvusException(1, "Compact auto-merge more than 60s")
> 
> [2022-06-23T19:54:14.107Z] E               pymilvus.exceptions.MilvusException: <MilvusException: (code=1, message=Compact auto-merge more than 60s)>
> ```

For this case only,  you could do the following to test auto compaction @binbinlv :
1. Don't load or index while auto compaction
2. Get query segment info after 30s.

**Comment by ThreadDao** [comment]
@XuanYang-cn 
1. Don't load or index while auto compaction

I guess users will definitely have such a scenario. We have no way to restrict users from not doing load and index during auto compact. It is also difficult for users to judge whether there is automatic compression

2. Get query segment info after 30s.
We allow the get_query_segment_info to be called many times within 60 seconds

**Comment by NicoYuan1986** [comment]
This issue no longer occurs in master(c58a8d8).
